nvflare/private/fed/server/fed_server.py

```diff
@@ -705,17 +686,16 @@ def overseer_callback(self, overseer_agent):
             return
 
         sp = overseer_agent.get_primary_sp()
-        # print(sp)
-        with self.engine.new_context() as fl_ctx:
-            self.server_state = self.server_state.handle_sd_callback(sp, fl_ctx)
+
+        with self.lock:
+            with self.engine.new_context() as fl_ctx:
+                self.server_state = self.server_state.handle_sd_callback(sp, fl_ctx)
 
         if isinstance(self.server_state, Cold2HotState):
-            server_thread = threading.Thread(target=self._turn_to_hot)
-            server_thread.start()
+            self._turn_to_hot()
```

Comment:
So now instead of running _turn_to_hot in a thread, you call it directly here. What could be the ramifications?
The turn_to_hot logic is complex and could take a while to finish. If running in a thread, then this overseer_callback will return quickly. This is potentially dangerous, since the CB could be called again by the overseer agent, which will create yet another thread to do state change.

So I think this is a good change to ensure that we only do one state change at a time.

Reply:
Yes, Yan you are very correct.

I observed that when running the old code, this function gets called multiple times.
And since this method also modify the state, it is like a bunch of threads keep modifying the state (from Cold to Cold2Hot, or from Cold2Hot to Hot).
So a racing condition there.

I agree it is better to ensure only do one state change at a time.

Reply:
There's a potential issue with using the lock here. The overseer calls the overseer_callback very quickly, the default is 6 seconds. Using lock could potentially cause multiple threads piling up waiting for execution.

Reply:
Good point!

In this case, the overseer_callback should simply return if it is currently working on state change. This means that we only work on one state change at a time.